qiskit/algorithms/optimizers/gradient_descent.py

```diff
@@ -12,17 +12,67 @@
 
 """A standard gradient descent optimizer."""
 
-from typing import Iterator, Optional, Union, Callable, Dict, Any, List, Tuple
-from functools import partial
-
+from dataclasses import dataclass, field
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List, Iterator
 import numpy as np
-
 from .optimizer import Optimizer, OptimizerSupportLevel, OptimizerResult, POINT
+from .steppable_optimizer import AskData, TellData, OptimizerState, SteppableOptimizer
+from .utils.learning_rate import LearningRate
 
 CALLBACK = Callable[[int, np.ndarray, float, float], None]
 
 
-class GradientDescent(Optimizer):
+@dataclass
+class GradientDescentState(OptimizerState):
+    """State of :class:`~.GradientDescent`.
+
+    Dataclass with all the information of an optimizer plus the learning_rate and the stepsize.
+    """
+
+    stepsize: Optional[float]
+    """Norm of the gradient on the last step."""
+    learning_rate: LearningRate = field(compare=False)
+    """Learing rate at the current step of the optimization process.
+
+    It behaves like a generator, (use ``next(learning_rate)`` to get the learning rate for the
+    next step) but it can also return  the current learning rate with ``learning_rate.current``.
+    """
+
+    def __init__(
+        self,
+        x: POINT,
+        fun: Optional[Callable[[POINT], float]],
+        jac: Optional[Callable[[POINT], POINT]],
+        nfev: Optional[int],
+        njev: Optional[int],
+        nit: Optional[int],
+        stepsize: Optional[float],
+        learning_rate: Union[float, List[float], np.ndarray, Callable[[], Iterator]],
+    ):
+        """
+        Args:
+            x: Current optimization parameters.
+            fun: Function being  optimized.
+            jac: Jacobian of the function being optimized.
+            nfev: Number of function evaluations so far in the optimization.
+            njev: Number of jacobian evaluations so far in the opimization.
+            nit: Number of optmization steps performed so far in the optimization.
+            stepsize: Norm of the jacobian on the last step.
+            learning_rate: A constant, list, array or factory of generators yielding learning rates
+            for the parameter updates. See the docstring for an example.
+        """
+
+        self.x = x
+        self.fun = fun
+        self.jac = jac
+        self.nfev = nfev
+        self.njev = njev
+        self.nit = nit
+        self.stepsize = stepsize
+        self.learning_rate = LearningRate(learning_rate)
```

Comment:
I have to wonder about this aspect. As a field its defined above as type `LearningRate` so this init is just for some simple conversion using which really the LearningRate constructor deals with details based on type. In the GradientDescent start() creates a LearningRate when it builds an instance of this state, which seems to contradict the types in the Union here which is presumably the set of type from which a LearningRate can be built. Now I do see in the unit test where it builds a GradientDescentState using an int. But I have to ask would it not be easier just have this type as LeaningRate here, get rid if this custom init and go with the default one that the dataclass will create (makes the code here simpler too as a result). This way it's always a LeaningRate and you don't need to worry about types here. If a user has some other types, then like the GradientDescent start method, it can simply create the LearningRate directly from that itself.

Reply:
You are right. The one that should be type LearningRate is the learning rate in the state of the system. self.learning_rate from the optimizer should only store the value from the user.
Thank you for spotting it.